### PR TITLE
Boyscout of repository and plugin related files

### PR DESCRIPTION
**Pull request by dbengsch**
(Should not appear in the changelog of the release)

**Boyscout:**

* Replace var by let or const
* Use class paradigm instead of constructor functions

Purpose: Prepares the refactor of the repositories in order to add the plugin repository.

### CODE REVIEW

File: lib/api/core/models/repositories/roleRepository.js [hunk 1]
```diff
@@ -8,212 +10,219 @@ var
...
+   *
+   * @param {Array} roleKeys The role ids to load
+   * @returns {Promise} Resolves to an array containing the matching found Role objects.
+   */
+  loadRoles (roleKeys) {
+    const
+      keys = [],
+      buffer = {},
+      result = [];
+    let role;
```
Comment (scottinet) [inline lib/api/core/models/repositories/roleRepository.js]: nitpicking: can be scoped inside the ` else if (this.kuzzle.config.security.standard.roles[roleKey])` block

**Comment by dbengsch** [comment]
@scottinet Done

File: lib/api/core/models/repositories/repository.js [hunk 2]
```diff
@@ -1,344 +1,358 @@
...
-    ttl = options.ttl;
-  }
-  if (ttl === false) {
-    return this.cacheEngine.set(key, JSON.stringify(this.serializeToCache(object)))
-      .then(() => object);
+  /**
+   * Delete repository from database according to its id
+   * @param id
+   */
+  deleteFromDatabase (id) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: nitpicking: why not adding an optional parameter `collection` here, as you did for `loadOneFromDatabase` method?

File: lib/api/core/models/repositories/repository.js [hunk 3]
```diff
@@ -1,344 +1,358 @@
...
+   * Persists the given ObjectConstructor object in cache.
+   * The opts optional parameters currently accept 2 options:
+   *   key: if provided, stores the object to the given key instead of the default one (<collection>/<id>)
+   *   ttl: if provided, overrides the default ttl set on the repository for the current operation.
+   *
+   * @param {object} object the object to persist
+   * @param {object} [opts] optional options for the current operation
+   * @returns {Promise}
+   */
+  persistToCache (object, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: nitpicking: add the default `opts` value within method declaration:
```
persistToCache(object, options = {}) {
// (...)
}
```

File: lib/api/core/models/repositories/repository.js [hunk 4]
```diff
@@ -1,344 +1,358 @@
...
+  /**
+   * Removes the object from the Cache Engine
+   * The opts optional parameters currently accepts only 1 option:
+   *   key: if provided, removes the given key instead of the default one (<collection>/<id>)
+   *
+   * @param {string} id
+   * @param {object} [opts={}] optional options for the current operation
+   * @returns {Promise}
+   */
+  deleteFromCache (id, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: see comment above.

File: lib/api/core/models/repositories/repository.js [hunk 5]
```diff
@@ -1,344 +1,358 @@
...
+    const
+      options = opts || {},
+      key = options.key || this.cacheKeyPrefix + id;
+
+    return this.cacheEngine.remove(key);
   }
 
-  return this.cacheEngine.expire(key, ttl);
-};
+  refreshCacheTTL (object, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: ditto

File: lib/api/core/models/repositories/repository.js [hunk 6]
```diff
@@ -1,344 +1,358 @@
...
+    const
+      options = opts || {},
+      key = options.key || this.cacheKeyPrefix + object._id;
+    let ttl = this.ttl;
```
Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: cant be
```javascript
const ttl = options.ttl || this.ttl;
```

Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: ok, forget it, it won't work if `options.ttl` is defined but `false`

File: lib/api/core/models/repositories/repository.js [hunk 7]
```diff
@@ -1,344 +1,358 @@
...
-/**
- * Serializes the object before being persisted to the database.
- *
- * @param {object} object The object to serialize
- * @returns {object}
- */
-Repository.prototype.serializeToDatabase = function repositorySerializeToDatabase (object) {
-  return object;
-};
+  expireFromCache (object, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/repository.js]: ditto

File: lib/api/core/models/repositories/profileRepository.js [hunk 8]
```diff
@@ -7,251 +9,252 @@ var
...
+    super(kuzzle);
+
+    this.collection = 'profiles';
+    this.ObjectConstructor = Profile;
+    this.profiles = {};
   }
 
-  if (typeof profileId !== 'string') {
-    return Promise.reject(new BadRequestError(`Invalid argument: Expected profile id to be a string, received "${typeof profileId}"`));
+  init () {
```
Comment (ballinette) [inline lib/api/core/models/repositories/profileRepository.js]: do we need to declare explicitly `init` method in each subclass if we declare it with a default empty parameter in the `Repository` class, like that:
```javascript
init(options = {}) {
// (...)
}
```

File: lib/api/core/models/repositories/tokenRepository.js [hunk 9]
```diff
@@ -8,94 +10,83 @@ var
...
-  this.ObjectConstructor = Token;
-  if (opts !== undefined && opts.ttl !== undefined) {
-    this.ttl = opts.ttl;
+class TokenRepository extends Repository {
+  /**
+   * @param {Kuzzle} kuzzle
+   * @param opts
+   * @constructor
+   */
+  constructor (kuzzle, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/tokenRepository.js]: niptpicking: 
```javascript
  constructor (kuzzle, opts = {}) {
    super(kuzzle);
    this.collection = 'token';
    this.ObjectConstructor = Token;
    if (opts.ttl !== undefined) {
      this.ttl = opts.ttl;
    }
}
```

File: lib/api/core/models/repositories/tokenRepository.js [hunk 10]
```diff
@@ -8,94 +10,83 @@ var
...
-    token = new Token(),
-    options = opts || {},
-    encodedToken,
-    error;
-
-  if (!user || user._id === null) {
-    error = new InternalError('Unknown User : cannot generate token');
-    return Promise.reject(error);
+  init () {
+    super.init({});
```
Comment (ballinette) [inline lib/api/core/models/repositories/tokenRepository.js]: Token should be stored only into the Cache, not into the DatabaseEngine.
So, as `Repository` subclass is using the internalStorage as default database engine, we should explicitly disable it at init:
```javascript
init() {
  super.init({databaseEngine: null});
}
```

File: lib/api/core/models/repositories/tokenRepository.js [hunk 11]
```diff
@@ -8,94 +10,83 @@ var
...
-  if (!options.expiresIn) {
-    options.expiresIn = this.kuzzle.config.security.jwt.expiresIn;
-  }
+  /**
+   * @param {User} user
+   * @param {Request} request
+   * @param {object} opts
+   * @returns {*}
+   */
+  generateToken (user, request, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/tokenRepository.js]: nitpicking: 
```javascript
generateToken (user, request, options = {}) {
  const token = new Token();
// (...)
}
```

File: lib/api/core/models/repositories/userRepository.js [hunk 12]
```diff
@@ -9,130 +11,134 @@ var
...
-
-  if (opts !== undefined && opts.ttl !== undefined) {
-    this.ttl = opts.ttl;
+class UserRepository extends Repository {
+  /**
+   * @param {Kuzzle} kuzzle
+   * @param opts
+   * @constructor
+   */
+  constructor (kuzzle, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/userRepository.js]: niptpicking: 
```javascript
constructor (kuzzle, opts = {}) {
  super(kuzzle);

  this.collection = 'users';
  this.ObjectConstructor = User;

  if (opts.ttl !== undefined) {
    this.ttl = opts.ttl;
  }
}
```

anyway, I wonder if the assignment of `this.ttl` should be moved to superclass. (I mean: for now, it's not implemented in `RoleRepository` and `ProfileRepository` classes, but maybe it should ?)

File: lib/api/core/models/repositories/userRepository.js [hunk 13]
```diff
@@ -9,130 +11,134 @@ var
...
 
-  return Repository.prototype.load.call(this, id)
-    .then(user => {
-      if (user === null) {
-        return null;
-      }
-      return this.hydrate(new User(), user);
-    });
-};
+  persist (user, opts) {
```
Comment (ballinette) [inline lib/api/core/models/repositories/userRepository.js]: nipticking: 
```javascript
persist(user, options = {}) {
// (...)
}
```

File: lib/api/core/models/repositories/userRepository.js [hunk 14]
```diff
@@ -9,130 +11,134 @@ var
...
 
-  // if the user exists (have an _id) but no profile
-  // set it to default
-  if (user.profileIds.length === 0) {
-    user.profileIds = ['default'];
-  }
+    // if the user exists (have an _id) but no profile
+    // set it to default
+    if (user.profileIds.length === 0) {
+      user.profileIds = ['default'];
```
Comment (ballinette) [inline lib/api/core/models/repositories/userRepository.js]: As for #727, we should get the default Profile Ids from `restrictedProfileIds` settings.

File: lib/api/core/models/repositories/repository.js (hunk 2 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/repository.js]: My bad, it was a try, I removed it in previous method.

File: lib/api/core/models/repositories/repository.js (hunk 3 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/repository.js]: Done in the next PR #723

File: lib/api/core/models/repositories/repository.js (hunk 4 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/repository.js]: Done in the next PR #723

File: lib/api/core/models/repositories/repository.js (hunk 5 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/repository.js]: Done in the next PR #723

File: lib/api/core/models/repositories/repository.js (hunk 7 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/repository.js]: Done in the next PR #723

File: lib/api/core/models/repositories/profileRepository.js (hunk 8 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/profileRepository.js]: the override is not used to define a default to options, but to force it to an empty object.

File: lib/api/core/models/repositories/tokenRepository.js (hunk 10 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/tokenRepository.js]: Completely right, I did something wrong here :(

File: lib/api/core/models/repositories/tokenRepository.js (hunk 11 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/tokenRepository.js]: Done in the next PR #723

File: lib/api/core/models/repositories/userRepository.js (hunk 12 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/userRepository.js]: @benoitvidis said me the RoleRepository and ProfileRepository was ment to force the disabling of Redis cache as they already have inmemory cache, but was somehow re-enabled by mistake, I'll create an issue about it to re-disable Redis cache.

File: lib/api/core/models/repositories/userRepository.js (hunk 13 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/userRepository.js]: Done in the next PR #723

File: lib/api/core/models/repositories/userRepository.js (hunk 14 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/userRepository.js]: Will be done in the next PR #723

File: lib/api/core/models/repositories/userRepository.js (hunk 12 above)
Comment (dbengsch) [inline lib/api/core/models/repositories/userRepository.js]: Issue #728

**Pull request merged**